File: 1653-number-of-good-leaf-nodes-pairs/number-of-good-leaf-nodes-pairs.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
#define MAX_LEAVES 1000

int count = 0;

int* dfs(struct TreeNode* node, int distance, int* size) {
    if (node == NULL) {
        *size = 0;
        return NULL;
    }

    if (node->left == NULL && node->right == NULL) {
        int* leaf = malloc(sizeof(int));
        leaf[0] = 0;
        *size = 1;
        return leaf;
    }

    int leftSize = 0, rightSize = 0;
    int* leftDistances = dfs(node->left, distance, &leftSize);
    int* rightDistances = dfs(node->right, distance, &rightSize);

    for (int i = 0; i < leftSize; i++) {
        for (int j = 0; j < rightSize; j++) {
            if (leftDistances[i] + rightDistances[j] + 2 <= distance) {
                count++;
            }
        }
    }

    int* result = malloc(MAX_LEAVES * sizeof(int));
    int resultSize = 0;

    for (int i = 0; i < leftSize; i++) {
        if (leftDistances[i] + 1 < distance) {
            result[resultSize++] = leftDistances[i] + 1;
        }
    }
    for (int i = 0; i < rightSize; i++) {
        if (rightDistances[i] + 1 < distance) {
            result[resultSize++] = rightDistances[i] + 1;
        }
    }

    free(leftDistances);
    free(rightDistances);

    *size = resultSize;
    return result;
}

int countPairs(struct TreeNode* root, int distance) {
    count = 0;
    int size;
    dfs(root, distance, &size);
    return count;
}

File: 1653-number-of-good-leaf-nodes-pairs/number-of-good-leaf-nodes-pairs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int count = 0;

    public int countPairs(TreeNode root, int distance) {
        dfs(root, distance);
        return count;
    }

    private List<Integer> dfs(TreeNode node, int distance) {
        if (node == null) {
            return new ArrayList<>();
        }

        if (node.left == null && node.right == null) {
            List<Integer> list = new ArrayList<>();
            list.add(0);
            return list;
        }

        List<Integer> leftDistances = dfs(node.left, distance);
        List<Integer> rightDistances = dfs(node.right, distance);

        for (int left : leftDistances) {
            for (int right : rightDistances) {
                if (left + right + 2 <= distance) {
                    count++;
                }
            }
        }

        List<Integer> result = new ArrayList<>();
        for (int left : leftDistances) {
            if (left + 1 < distance) {
                result.add(left + 1);
            }
        }
        for (int right : rightDistances) {
            if (right + 1 < distance) {
                result.add(right + 1);
            }
        }

        return result;
    }
}